src/object_store/cache.rs

```diff
@@ -217,88 +223,201 @@ impl CachingObjectStore {
+                .entry_by_ref(&key)
+                .or_insert(CacheValue::Pending(rx.clone()))
+                .await;
+
+            let is_fresh = entry.is_fresh();
+            let mut chunk_data = None;
+            if !is_fresh {
+                chunk_data = Some(match entry.into_value() {
+                    CacheValue::Pending(mut rx) =>
+                    // Since the entry is not fresh another task is evaluating this future,
+                    // so wait on it.
+                    // Also use a timeout to break from the source task potentially hanging/deadlocking.
+                    {
+                        match tokio::time::timeout(Duration::from_secs(5), rx.changed())
+                            .await
+                        {
+                            Ok(maybe_receive) => {
+                                maybe_receive.map_err(|e| {
+                                    object_store::Error::Generic {
+                                        store: "cache_store",
+                                        source: Box::new(e),
+                                    }
+                                })?;
+                                debug!("{key:?} awaited");
+                                rx.borrow().clone().map_err(|e| {
+                                    object_store::Error::Generic {
+                                        store: "cache_store",
+                                        source: Box::new(e),
+                                    }
+                                })?
+                            }
+                            Err(_) => {
+                                debug!("{key:?} refetching");
+                                // TODO: maybe check that value is not in cache now,
+                                // and if not insert + trigger persisting to file
+                                let data = self
+                                    .inner
+                                    .get_range(path, chunk_range.clone())
+                                    .await?;
+                                debug!("{key:?} refetched");
+                                data
+                            }
                         }
                     }
+                    CacheValue::Memory(data) => {
+                        debug!("Cache value for {key:?} fetched from memory");
+                        data.clone()
+                    }
+                    CacheValue::File(path, _) => {
+                        debug!("Cache value for {key:?} fetched from file");
+                        self.file_manager.read_file(path).await.unwrap()
+                    }
                 });
+            } else {
+                // Entry is fresh; we need to evaluate it in the upcoming batch...
+                chunk_batches.push((key.clone(), tx));
+                // ... and also create a new sender/receiver for the next pending entry
+                (tx, rx) = channel(Ok(Bytes::new()));
```

Comment:
We end up with one orphan channel here at the end that's never used, right? (since we create one in the beginning, assign it if we need to do a fetch + create a new one for the next fetch)

If it's one channel per chunk, do we hit weird scalability issues if we're trying to download a few thousand 1MB ranges?

Reply:
Correct on the orphan channel comment, but it would add even more complexity in trying to avoid it.

Not sure about scalability issues, haven't tested it against anything more involved than TPC-H SF10